File: Python/110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def getDepth(self, root: TreeNode) -> int:
        if root == None:
            return 0
        lDepth = self.getDepth(root.left) #左子树高度
        rDepth = self.getDepth(root.right) #右子树高度
        #左右子树不符合要求, 或者高度差超过1
        if lDepth < 0 or rDepth < 0 or abs(lDepth - rDepth) > 1:
            return -1
        else:
            return max(lDepth, rDepth) + 1
    def isBalanced(self, root: TreeNode) -> bool:
        if root == None:
            return True
        lDepth = self.getDepth(root.left) #左子树高度
        rDepth = self.getDepth(root.right) #右子树高度
        #左右子树不符合要求, 或者高度差超过1
        if lDepth < 0 or rDepth < 0 or abs(lDepth - rDepth) > 1:
            return False
        else:
            return True

File: Go/110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func abs(x int) int {
    if x >= 0 {
        return x
    } else {
        return -x
    }
}

func max(x, y int) int {
    if x > y {
        return x
    } else {
        return y
    }
}

func getDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    //取得左右子树的高度
    lDepth, rDepth := getDepth(root.Left), getDepth(root.Right)
    //左右子树不符合要求, 或者它们的高度差不符合要求
    if lDepth < 0 || rDepth < 0 || abs(lDepth - rDepth) > 1 {
        return -1
    }
    return max(lDepth, rDepth) + 1
}

func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true
    }
     //取得左右子树的高度
    lDepth, rDepth := getDepth(root.Left), getDepth(root.Right)
    //左右子树不符合要求, 或者它们的高度差不符合要求
    if lDepth < 0 || rDepth < 0 || abs(lDepth - rDepth) > 1 {
        return false
    }
    return true
}